buildroot/share/scripts/upload.py

```diff
@@ -0,0 +1,193 @@
+    upload_reset = True                             # Trigger a soft reset for firmware update after the upload
+
+    if Debug:
+        print("Upload using:")
+        print(f" -Source      : {upload_firmware_source_name}")
+        print(f" -Target      : {upload_firmware_target_name}")
+        print(f" -Port        : {upload_port} @ {upload_speed} baudrate")
+        print(f" -Timeout     : {upload_timeout}")
+        print(f" -Block size  : {upload_blocksize}")
+        print(f" -Compression : {upload_compression}")
+        print(f" -Error ratio : {upload_error_ratio}")
+        print(f" -Test        : {upload_test}")
+        print(f" -Reset       : {upload_reset}")
+        print("----------------")
+
+    # Init serial port
+    port = serial.Serial(upload_port, baudrate = upload_speed, write_timeout = 0, timeout = 0.1)
+    port.reset_input_buffer()
+
+    # Check SD card status
+    if not _CheckSDCard(): exit(2)
+    
+    # Get firmware files
+    (Status, FirmwareFiles) = _GetFirmwareFiles()
+    if not Status: exit(3)
+    if Debug:
+        for FirmwareFile in FirmwareFiles:
+            print("Found: " + FirmwareFile)
+        
+    # Get all 1st level firmware files (to remove)
+    OldFirmwareFiles = _FilterFirmwareFiles(FirmwareFiles[1:len(FirmwareFiles)-2])   # Skip header and footers of list
+    if len(OldFirmwareFiles) == 0:
+        print("No old firmware files found")
+    else:
+        print(f"Remove {len(OldFirmwareFiles)} old firmware file{'s' if len(OldFirmwareFiles) != 1 else ''}:")
+        for OldFirmwareFile in OldFirmwareFiles:
+            print(" -Removing- " + OldFirmwareFile + "...", end='', flush=True)
+            print("OK" if _RemoveFirmwareFile(OldFirmwareFile) else "Error!")
+
+    # Cleanup completed
+    port.close()
+    if Debug: print("Cleanup completed")
+    
+    try:
+        # Upload File
+        print("Copying '" + upload_firmware_source_name + "' --> '" + upload_firmware_target_name + "'")
+        protocol = MarlinBinaryProtocol.Protocol(upload_port, upload_speed, upload_blocksize, float(upload_error_ratio), int(upload_timeout))
+        #echologger = MarlinBinaryProtocol.EchoProtocol(protocol)
+        protocol.connect()
+        filetransfer = MarlinBinaryProtocol.FileTransferProtocol(protocol)
+        filetransfer.copy(upload_firmware_source_name, upload_firmware_target_name, upload_compression, upload_test)
+        protocol.disconnect()
+
+        # Notify upload completed
+        protocol.send_ascii("M117 Firmware uploaded")
+
+        # Remount SD card
+        print('Wait for SD card release...')
+        time.sleep(1)
+        print('Remount SD card')
+        protocol.send_ascii("M21")
+
+        if upload_reset:
+            print("Trigger firmware update...")
+            protocol.send_ascii("M997", True)
+    except KeyboardInterrupt:
+        if filetransfer: filetransfer.abort()
+
+    except MarlinBinaryProtocol.FatalError:
+        print("Too Many Retries, Abort")
```

Comment:
In these cases we would want PlatformIO to report the upload as a failure, but I don't see anything that would cause that to happen.

Perhaps these need to re-raise the caught exceptions, so that PIO will know something went wrong? There could also be some other way this is communicated.

Reply:
I see.
This part surely need improvements.
Thanks